zebra/zebra_rib.c

```diff
@@ -3235,6 +3235,32 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 
 	same = first_same;
 
+	if (!startup &&
+	    (re->flags & ZEBRA_FLAG_SELFROUTE) && zrouter.asic_offloaded) {
+		if (!same) {
+			if (IS_ZEBRA_DEBUG_RIB)
+				zlog_debug("prefix: %pRN is a self route where we do not have an entry for it.  Dropping this update, it's useless", rn);
+			/*
+			 * We are not on startup, this is a self route
+			 * and we have asic offload.  Which means
+			 * we are getting a callback for a entry
+			 * that was already deleted to the kernel
+			 * but an earlier response was just handed
+			 * back.  Drop it on the floor
+			 */
+			if (re->nhe && re->nhe_id) {
```

Comment:
I'm a little concerned about duplicating this block of cleanup code: I'm not sure even the version in rib_unlink() is really complete. Could we maybe have a rib_re_nhg_free() api that has the core version of this, and make sure we clean up backup nexthops, and the opaque struct?

Reply:
fixed the duplication.  I'm not sure the proper way to cleanup the backup nexthops and opaque struct is not created in this code path.

Reply:
thanks for adding the helper api; I'll circle back around once it's merged and see about including more of the re stuff